common/pose.h

```diff
@@ -1,23 +1,209 @@
 
-//! A generic template for 3D unit arrays
-template<class T>
-using Vector3 = std::array<T, 3>;
+//! A generic template for 2D arrays
+template<typename T>
+using Array2 = std::array<T, 2>;
+
+//! A generic template for 3D arrays
+template<typename T>
+using Array3 = std::array<T, 3>;
+
+//! Base class for vectors of length units
+template<typename LengthUnit, size_t N>
+class VectorBase
+{
+    static_assert(units::traits::is_length_unit<LengthUnit>::value,
+                  "LengthUnit is not a unit of length");
+
+public:
+    VectorBase() = default;
+
+    template<typename... Ts>
+    VectorBase(Ts &&... args)
+      : m_Array({ std::forward<Ts>(args)... })
+    {}
+
+    operator const std::array<LengthUnit, N> &() const
+    {
+        return m_Array;
+    }
+
+    LengthUnit &operator[](size_t i) { return m_Array[i]; }
+    const LengthUnit &operator[](size_t i) const { return m_Array[i]; }
+    static constexpr size_t size() { return N; }
+
+    auto begin() { return m_Array.begin(); }
+    auto begin() const { return m_Array.begin(); }
+    auto end() { return m_Array.end(); }
+    auto end() const { return m_Array.end(); }
+    auto cbegin() const { return m_Array.cbegin(); }
+    auto cend() const { return m_Array.end(); }
+
+private:
+    std::array<LengthUnit, N> m_Array;
+};
+
+template<typename LengthUnit>
+class Vector3;
+
+//! 2D length unit vector
+template<typename LengthUnit>
+class Vector2
+  : public VectorBase<LengthUnit, 2>
+{
+public:
+    Vector2() = default;
+
+    Vector2(LengthUnit x, LengthUnit y)
+      : VectorBase<LengthUnit, 2>(x, y)
+    {}
+
+    Vector2(const std::array<LengthUnit, 2> &array)
+      : Vector2(array[0], array[1])
+    {}
+
+    operator Vector3<LengthUnit>() const { return Vector3<LengthUnit>(x(), y(), z()); }
+
+    LengthUnit &x() { return (*this)[0]; }
+    const LengthUnit &x() const { return (*this)[0]; }
+    LengthUnit &y() { return (*this)[1]; }
+    const LengthUnit &y() const { return (*this)[1]; }
+    static constexpr LengthUnit z() { return LengthUnit(0); }
+};
+
+//! 3D length unit vector
+template<typename LengthUnit>
+class Vector3
+  : public VectorBase<LengthUnit, 3>
+{
+public:
+    Vector3() = default;
+
+    Vector3(LengthUnit x, LengthUnit y, LengthUnit z)
+      : VectorBase<LengthUnit, 3>(x, y, z)
+    {}
+
+    Vector3(const std::array<LengthUnit, 3> &array)
+      : Vector3(array[0], array[1], array[2])
+    {}
+
+    operator Vector2<LengthUnit>() const { return Vector2<LengthUnit>(x(), y()); }
+
+    LengthUnit &x() { return (*this)[0]; }
+    const LengthUnit &x() const { return (*this)[0]; }
+    LengthUnit &y() { return (*this)[1]; }
+    const LengthUnit &y() const { return (*this)[1]; }
+    LengthUnit &z() { return (*this)[2]; }
+    const LengthUnit &z() const { return (*this)[2]; }
+};
+
+template<typename LengthUnit, typename AngleUnit>
+class Pose3;
+
+//! A two-dimensional pose
+template<typename LengthUnit, typename AngleUnit>
+class Pose2
```

Comment:
I think deriving off anything std::anything is generally viewed as a bad idea and, in this case, there's particularly little point as according to [this](https://stackoverflow.com/questions/37186443/confusion-while-deriving-from-stdtuple-can-not-handle-stdget) the cruddy ``std::get<n>(x)`` stuff doesn't even survive inheritance. Why not just have the length and angles as members? I also think ``return m_Position.x()`` looks way nicer than ``return std::get<0>(*this)[0];``